### PR TITLE
Follow up optimizations to getAllItems() and getItems()

### PR DESCRIPTION
**Pull request by res0nance**
There was a missed optimization opportunity in #4469 to do filtering in child items automatically.

### Proposed changelog entries

* Entry 1: Use new API to optimize getAllItems() and getItems() calls
* ...

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] JIRA issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`

### CODE REVIEW

**Review by varyvol — APPROVED** [review]

**Review by timja — APPROVED** [review]

**Review by markjacksonfishing — APPROVED** [review]

File: core/src/main/java/hudson/model/Items.java [hunk 1]
```diff
@@ -421,7 +421,7 @@ public static XmlFile getConfigFile(Item item) {
         return r;
     }
     private static <T extends Item> void getAllItems(final ItemGroup root, Class<T> type, List<T> r, Predicate<T> pred) {
-        List<Item> items = new ArrayList<>(((ItemGroup<?>) root).getItems());
+        List<Item> items = new ArrayList<>(((ItemGroup<?>) root).getItems(t -> t instanceof ItemGroup || (type.isInstance(t) && pred.test(type.cast(t)))));
         // because we add items depth first, we can use the quicker BY_NAME comparison
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Items.java]: Do we need all checks below after adding these predicates?

Comment (res0nance) [reply core/src/main/java/hudson/model/Items.java]: Yes there is a sneaky edge case where type is `ItemGroup`. Now that I think about it I've forgotten to add a test for this and will be adding one.

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 2]
```diff
@@ -1796,9 +1796,8 @@ public String getFullDisplayName() {
      */
     public <T> List<T> getItems(Class<T> type) {
         List<T> r = new ArrayList<>();
-        for (TopLevelItem i : getItems())
-            if (type.isInstance(i))
-                 r.add(type.cast(i));
+        for (TopLevelItem i : getItems(type::isInstance))
```
Comment (alecharp) [inline core/src/main/java/jenkins/model/Jenkins.java]:
```suggestion
        for (TopLevelItem i : getItems(type::isInstance)) {
```

**Review by fcojfernandez — APPROVED** [review]

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 3]
```diff
@@ -1796,9 +1796,8 @@ public String getFullDisplayName() {
...
+             r.add(type.cast(i));
```
Comment (alecharp) [inline core/src/main/java/jenkins/model/Jenkins.java]:
```suggestion
             r.add(type.cast(i));
         }
```

**Review by alecharp — APPROVED** [review]
Seems good even if I would prefer the usage of `{` / `}` in loops.

**Comment by timja** [comment]
Thanks for the PR! Last call for any feedback, it will be merged after ~24 hours if there's no negative feedback
